Kernel/FileSystem/RAMFS/Inode.cpp

```diff
@@ -97,32 +97,66 @@ ErrorOr<NonnullOwnPtr<RAMFSInode::DataBlock>> RAMFSInode::DataBlock::create()
     return TRY(adopt_nonnull_own_or_enomem(new (nothrow) DataBlock(move(data_block_buffer_vmobject))));
 }
 
-ErrorOr<void> RAMFSInode::ensure_allocated_blocks(size_t offset, size_t io_size)
+ErrorOr<void> RAMFSInode::increase_blocks_count(size_t last_block_index)
 {
     VERIFY(m_inode_lock.is_locked());
-    size_t block_start_index = offset / DataBlock::block_size;
-    size_t block_last_index = ((offset + io_size) / DataBlock::block_size) + (((offset + io_size) % DataBlock::block_size) == 0 ? 0 : 1);
-    VERIFY(block_start_index <= block_last_index);
+    VERIFY(last_block_index > m_blocks.size());
+    TRY(fs().current_storage_usage_size({}).with_exclusive([&](auto& current_storage_usage_size) -> ErrorOr<void> {
+        if (Checked<u64>::multiplication_would_overflow(static_cast<u64>(last_block_index - m_blocks.size()), DataBlock::block_size))
+            return Error::from_errno(EOVERFLOW);
+
+        u64 storage_diff = static_cast<u64>(last_block_index - m_blocks.size()) * DataBlock::block_size;
+        if (Checked<u64>::addition_would_overflow(current_storage_usage_size, storage_diff))
+            return Error::from_errno(EOVERFLOW);
+        if (fs().max_size().has_value()) {
+            if (Checked<u64>::addition_would_overflow(fs().max_size().value(), current_storage_usage_size + storage_diff))
+                return Error::from_errno(EOVERFLOW);
+            if (fs().max_size().value() < (current_storage_usage_size + storage_diff)) {
+                VERIFY((fs().max_size().value() % DataBlock::block_size) == 0);
+                return Error::from_errno(ENOSPC);
+            }
+        }
 
-    size_t original_size = m_blocks.size();
-    Vector<size_t> allocated_block_indices;
-    ArmedScopeGuard clean_allocated_blocks_on_failure([&] {
-        for (auto index : allocated_block_indices)
-            m_blocks[index].clear();
-        MUST(m_blocks.try_resize(original_size));
+        TRY(m_blocks.try_resize(last_block_index));
```

Comment:
I do `VERIFY(last_block_index > m_blocks.size());` so I think this is OK?
Or maybe I didn't understand what you meant here...

Reply:
If the variable name is accurate, then let's assume something needs one block of space, therefore the last block index of the respective function call would be 0. Then `m_blocks.try_resize(0)` gives the wrong result, since we need one block of space, not zero.